File: hackerrank/problem solving/data structures/1 - arrays/3 - dynamic array/dynamic_array.rb
# https://www.hackerrank.com/challenges/dynamic-array/problem?isFullScreen=true

def main
  n, n_queries = read_numbers
  obj = DynamicArray.new(n)
  n_queries.times { obj.handle_query(read_numbers) }
  puts obj.answers
end

def read_numbers
  gets.split.map!(&:to_i)
end

class DynamicArray
  attr_reader :answers

  def initialize(n)
    @n = n
    @arr = Array.new(n).map { [] }
    @answers = []
    @last_answer = 0
  end

  def handle_query(query)
    type, x, y = query
    index = (x ^ @last_answer) % @n

    if type == 1
      @arr[index] << y
    else
      j = y % @arr[index].size
      @last_answer = @arr[index][j]
      @answers.push(@last_answer)
    end
  end
end

main if __FILE__ == $0

File: hackerrank/problem solving/data structures/1 - arrays/3 - dynamic array/dynamic_array.py
# https://www.hackerrank.com/challenges/dynamic-array/problem?isFullScreen=true

def main():
    n, n_queries = read_numbers()
    obj = DynamicArray(n)

    for _ in range(n_queries):
        obj.handle_query(read_numbers())

    print(*obj.answers(), sep='\n')


def read_numbers():
    return list(map(int, input().split()))


class DynamicArray:

    def __init__(self, n):
        self._n = n
        self._arr = [[] for _ in range(n)]
        self._answers = []
        self._last_answer = 0

    def handle_query(self, query):
        type, x, y = query
        index = (x ^ self._last_answer) % self._n

        if type == 1:
            self._arr[index].append(y)
        else:
            j = y % len(self._arr[index])
            self._last_answer = self._arr[index][j]
            self._answers.append(self._last_answer)

    def answers(self):
        return self._answers


if __name__ == '__main__':
    main()
